rts/lra/lra-test/lra-test-arquillian-extension/src/main/java/io/narayana/lra/arquillian/spi/NarayanaLraRecovery.java

```diff
@@ -0,0 +1,57 @@
+package io.narayana.lra.arquillian.spi;
+
+import io.narayana.lra.LRAConstants;
+import io.narayana.lra.client.NarayanaLRAClient;
+import org.eclipse.microprofile.lra.tck.service.spi.LraRecoveryService;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.client.WebTarget;
+import javax.ws.rs.core.Response;
+import java.net.URI;
+
+@ApplicationScoped
+public class NarayanaLraRecovery implements LraRecoveryService {
+
+    @Override
+    public void triggerRecovery() {
+        doTriggerRecovery(System.getProperty(NarayanaLRAClient.LRA_COORDINATOR_HOST_KEY),
+            Integer.parseInt(System.getProperty(NarayanaLRAClient.LRA_COORDINATOR_PORT_KEY)));
+    }
+
+    @Override
+    public void triggerRecovery(URI lraId) {
+        doTriggerRecovery(lraId.getHost(), lraId.getPort());
+    }
+
+    private void doTriggerRecovery(String host, int port) {
+        if (!invokeCoordinatorRecovery(host, port)) {
+            // first recovery scan probably collided with periodic recovevery which started
+            // before the test execution so try once more
+            invokeCoordinatorRecovery(host, port);
+        }
```

Comment:
The two recovery attempts could be due to this? /cc @xstefank @mmusgrov 

Perhaps there is something in the json payload below even in the case we expect and that might cause forget twice (PS I am not sure if forget should be called twice even in two recovery attempts - @mmusgrov might be able to comment on that - perhaps after forget is called once Narayana should remember that and in the absence of a crash it could save that info)

Reply:
yes, it is because of this. But recovery should forget about LRA once forget call succeeds for the first time. I have debug logs locally on the coordinator, it seems that the same LRA is registered for recovery twice with a different id. Need to look a little more.

Reply:
Very strange. 

I do note that the console output shows the have the same LRA ID (I think) but perhaps this is not the ID you are referring to?:
> [INFO] [lra-coordinator-it-test] 2019-10-02 22:53:00,405 INFO  [io.nar.lra] (executor-thread-1) RecoveringLRA.replayPhase2 for http://localhost:8080/lra-coordinator/0_ffffac11000c_81fb_5d952910_6d ended with status: Closing
> [INFO] [lra-coordinator-it-test] 2019-10-02 22:53:05,440 INFO  [io.nar.lra] (executor-thread-1) RecoveringLRA.replayPhase2 for http://localhost:8080/lra-coordinator/0_ffffac11000c_81fb_5d952910_6d ended with status: Closing

Reply:
no, this is correct. triggerRecovery() is called two times in the test. My other comments about duplicate id were a dead end (different issue - quarkus is now by default creating objectstore on the filesystem so it's not just enough to kill coordinator process now to clear out LRAs). 

The issue is indeed in how to determine that we collided with periodic recovery (and thus need to replay recovery call). The issue with testForget is:

* test completes an LRA with participant returning 503 status
* triggerRecovery is called
** first /recovery call calls participant /status which again return 503 (thus LRA is still active)
** now, because the /recovery returned non-empty JSON we will replay the /recovery call (this error) -- HERE we need to know that we do not need to replay this call
* then the test changes participant return statuses to 200
* triggerRecovery is called again
* participant /status is called which now returns 200
* participant /forget is called which returns 200

@mmusgrov @tomjenkinson is there a way we can tell if recovery scan is running?

Note that there is nothing wrong with this behavior (status can be called multiple times) but the test is not adjusted to it.

Reply:
@xstefank I didn't quite follow the quarkus issue. Isn't the objectstore always on disk? If there is some problem there @mmusgrov please take a look. To the case in hand, regarding knowing if the recovery scan is in progress I think we should be OK to use something like: https://github.com/jbosstm/narayana/blob/master/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/recovery/RecoveryManager.java#L256 and https://github.com/jbosstm/narayana/blob/master/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/recovery/RecoveryManager.java#L282 - or perhaps we can start the recovery manager in non-threaded mode if we  have the callbacks we need in order to trigger recovery: https://github.com/jbosstm/narayana/blob/master/ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/RecoveryManagerImple.java#L63

Reply:
I am not really following the issue here but:
- calling forget multiple times is within spec (for example suppose the participant gets the notification but the response goes missing;
- calling scan will block if there is a current scan in progress and if the record wasn't recovered by that scan then calling scan a second time will recover it completely (if the participants can be contacted).

Will you clearly state what the issue is again?

Reply:
@mmusgrov First issue:

if the /recovery call collides with the periodic recovery then LRAs ended after periodic recovery started but before /recovery call is received are not recovered as /recovery call is just waiting for the periodic recovery scan to finish
* solved by calling /recovery twice

Second issue:

in the test `testForget` the **`/status`** is called twice (not /forget!) because of the test scenario which is described in detail in https://github.com/jbosstm/narayana/pull/1510#discussion_r330947731

Reply:
So I still do not see any issue, that is all expected behaviour. The PR looks fine as is.

Reply:
With one caveat: why doesn't the test trigger the second recovery invocation. The typical behaviour would be for the TCK:
- to test a condition and trigger recovery if required
- retest the condition and rerun recovery if required

If you don't to burden the code with this responsibility then an possible alternative would be for the TCK test to pass in a lamda to validate the condition.

With the way you have coded it any test that fails with pending LRAs will now fail.